lib/Sema/TypeCheckAttr.cpp

```diff
@@ -5475,4 +5475,35 @@ void AttributeChecker::visitReasyncAttr(ReasyncAttr *attr) {
   // FIXME
 }
 
-void AttributeChecker::visitAtReasyncAttr(AtReasyncAttr *attr) {}
\ No newline at end of file
+void AttributeChecker::visitAtReasyncAttr(AtReasyncAttr *attr) {}
+
+namespace {
+
+class ClosureAttributeChecker
+    : public AttributeVisitor<ClosureAttributeChecker> {
+  ASTContext &ctx;
+public:
+  ClosureAttributeChecker(ClosureExpr *closure)
+    : ctx(closure->getASTContext()) { }
+
+  void visitDeclAttribute(DeclAttribute *attr) {
+    ctx.Diags.diagnose(
+        attr->getLocation(), diag::unsupported_closure_attr,
+        attr->isDeclModifier(), attr->getAttrName())
+      .fixItRemove(attr->getRangeWithAt());
+    attr->setInvalid();
+  }
+
+  void visitConcurrentAttr(ConcurrentAttr *attr) {
+    // Nothing else to check.
```

Comment:
Do we also need to check if the parameter types and result type are ConcurrentValue, or is that not implemented for functions yet either?

Reply:
We've determined that we don't need that checking, because the act of calling a`@concurrent` function doesn't necessarily make it share its arguments or result value across different concurrency domains. It's the actual capture of non-concurrent values that is the problem.